### PR TITLE
feat(mssql): mssql extension for query lineage parsing

### PR DESCRIPTION
**Pull request by alokr-dhub**
## Summary

Adds query-based lineage and usage statistics extraction for MS SQL Serve
**Query History Lineage Extraction**:
- ✅ Query Store support (SQL Server 2016+) - primary method
- ✅ DMV fallback (`sys.dm_exec_cached_plans`, `sys.dm_exec_sql_text`) - legacy support
- ✅ Automatic method detection with graceful fallback
- ✅ Usage statistics: query frequency, execution time, user attribution
- ✅ Parsed lineage from SQL queries (INSERT INTO SELECT, CTAS, etc.)

## Files Changed
**New Files**:
- `lineage.py`: Query extraction and lineage population logic
- `query.py`: SQL query templates for Query Store and DMV
- `alias_filter.py`: TSQL alias filtering for procedure lineage
- `test_mssql_query_lineage_integration.py`: Integration tests

**Modified Files**:
- `source.py`: Integration with SQL parsing aggregator, config validation
- `mssql_pre.md`: User-facing documentation
- `mssql_recipe.yml`: Configuration examples

## Testing
- ✅ Unit tests for query extraction, validation, and error handling
- ✅ Integration tests against real MSSQL with Query Store
- ✅ Edge case coverage (permissions, version mismatches, Unicode)
- ✅ No linter errors

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [PR Title Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#pr-title-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py [hunk 1]
```diff
@@ -268,6 +407,29 @@ def __init__(
...
+
+            self.sql_aggregator = SqlParsingAggregator(
+                platform=self.platform,
+                platform_instance=self.config.platform_instance,
+                env=self.config.env,
+                graph=self.ctx.graph,
+                generate_lineage=True,
+                generate_queries=True,
+                generate_usage_statistics=self.config.include_usage_statistics,
+                usage_config=self.config  # type: ignore[arg-type]
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py]: In line 395, self.config is SQLServerConfig type. However, this SqlParsingQueryAggregator expects usage_config based on BaseUsageConfig type. Is it OK? I'm not sure if we want to assign this with type-ignore.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    def check_dmv_permissions() -> TextClause:
+        """Check if user has VIEW SERVER STATE permission for DMVs."""
+        return text("""
+            SELECT 
+                HAS_PERMS_BY_NAME(NULL, NULL, 'VIEW SERVER STATE') AS has_view_server_state
+        """)
+
+    @staticmethod
+    def get_query_history_from_query_store(
+        database: Optional[str],
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py]: this variable is not being used inside this function. Do we need this?

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            HAVING SUM(rs.count_executions) >= :min_calls
+            ORDER BY SUM(rs.avg_duration * rs.count_executions) DESC
+        """
+        )
+
+        return MSSQLQuery._finalize_query(query, exclude_patterns, limit, min_calls)
+
+    @staticmethod
+    def get_query_history_from_dmv(
+        database: Optional[str],
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py]: This variable is never used in this function.

File: metadata-ingestion/docs/sources/mssql/mssql_pre.md [hunk 4]
```diff
@@ -49,3 +71,380 @@ If you encounter permission errors:
...
+#### Enable Query Store
+
+```sql
+-- Enable Query Store for your database
+ALTER DATABASE [YourDatabase] SET QUERY_STORE = ON;
+
+-- Configure Query Store settings (recommended)
+ALTER DATABASE [YourDatabase] SET QUERY_STORE (
+    OPERATION_MODE = READ_WRITE,
+    DATA_FLUSH_INTERVAL_SECONDS = 900,
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/docs/sources/mssql/mssql_pre.md]: Just out of curiosity, where are these numbers coming from?

File: metadata-ingestion/docs/sources/mssql/mssql_pre.md [hunk 5]
```diff
@@ -49,3 +71,380 @@ If you encounter permission errors:
...
+   - Requires Query Store to be enabled
+
+2. **DMV Fallback** - All supported versions
+   - Uses `sys.dm_exec_cached_plans` and related DMVs
+   - Limited to queries currently in plan cache
+   - Smaller query history window
+
+The source will automatically detect and use the appropriate method based on your SQL Server version and configuration.
+
+### Troubleshooting
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/docs/sources/mssql/mssql_pre.md:240]: Nice to have this one.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
+import logging
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:1]: This lineage.py is for query level lineage. Can we rename it into something like query_lineage_extractor.py? That would be more straight forward.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+    """Represents a single query entry from MS SQL Server query history."""
+
+    query_id: str
+    query_text: str
+    execution_count: int
+    total_exec_time_ms: float
+    database_name: str
+
+    @property
+    def avg_exec_time_ms(self) -> float:
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py]: This property is not being used in prod level. Do we really need this?

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+
+        logger.info("Query Store is not enabled - falling back to DMV-based extraction")
+        return False
+
+    def _check_dmv_permissions(self) -> bool:
+        """Check if user has VIEW SERVER STATE permission for DMVs."""
+        result = self.connection.execute(MSSQLQuery.check_dmv_permissions())
+        row = result.fetchone()
+
+        if not row or not row["has_view_server_state"]:
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py]: The way to check it's available is different from the above "_check_query_store_available" method. It would be good to be consistent.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+            logger.error(
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py]: "Warning" level is good enough.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+            return False, f"Unexpected permission check failure: {e}", "none"
+
+        return True, "DMV-based extraction available", "dmv"
+
+    def extract_query_history(self) -> list[MSSQLQueryEntry]:
+        """Extract queries using the best available method (Query Store or DMVs)."""
+        is_ready, message, method = self.check_prerequisites()
+
+        if not is_ready:
+            logger.error(
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py]: Warning level is good enough here.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+    def extract_query_history(self) -> list[MSSQLQueryEntry]:
+        """Extract queries using the best available method (Query Store or DMVs)."""
+        is_ready, message, method = self.check_prerequisites()
+
+        if not is_ready:
+            logger.error(
+                f"Query history extraction not available: {message}. "
+                "Query-based lineage will be skipped."
+            )
+            self.report.report_failure(
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py]: This isn't a failure, it's a limitation.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py [hunk 12]
```diff
@@ -0,0 +1,335 @@
...
+        if not row or not row["has_view_server_state"]:
+            logger.error(
+                "Insufficient permissions. Grant VIEW SERVER STATE permission: "
+                "GRANT VIEW SERVER STATE TO [datahub_user];"
+            )
+            return False
+
+        return True
+
+    def check_prerequisites(self) -> tuple[bool, str, str]:
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py]: Return type would be better to be NameTuple or a dataclass. Also, enum type is good for the last element.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 13]
```diff
@@ -0,0 +1,180 @@
+import logging
+from typing import List, Optional, Tuple
+
+from sqlalchemy import text
+from sqlalchemy.sql.elements import TextClause
+
+logger = logging.getLogger(__name__)
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py]: This logger never used.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+        SECURITY: column_expr is ALWAYS a hardcoded column name from this module
+        (never user input). Pattern values use bind parameters (:exclude_0, etc.)
+        which are safely bound by SQLAlchemy.
+        """
+        if not exclude_patterns:
+            return ""
+
+        conditions = []
+        for i in range(len(exclude_patterns)):
+            condition = column_expr + " NOT LIKE :exclude_" + str(i)
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py]: Can we use f-string?
```
condition = f"{column_expr} NOT LIKE :exclude_{i}"
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+    ) -> dict:
+        """
+        Build parameter dict with exclude patterns (exclude_0, exclude_1, etc.).
+
+        These params are bound by SQLAlchemy's parameterization system, preventing SQL injection.
+        """
+        params = base_params.copy()
+        if exclude_patterns:
+            for i, pattern in enumerate(exclude_patterns):
+                key = "exclude_" + str(i)
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py]: Would be good to use f-string
```
key = f"exclude_{i}"
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py [hunk 16]
```diff
@@ -0,0 +1,335 @@
...
+            )
+        else:  # dmv
+            query, params = MSSQLQuery.get_query_history_from_dmv(
+                database=self.config.database,
+                limit=self.config.max_queries_to_extract,
+                min_calls=self.config.min_query_calls,
+                exclude_patterns=self.config.query_exclude_patterns,
+            )
+
+        with PerfTimer() as timer:
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:228]: QQ: What's the purpose of this measurement? Do we want to check bottleneck? There are 3 timer for logging. Would be good to know what's the purpose.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py [hunk 17]
```diff
@@ -0,0 +1,335 @@
...
+                self.report.report_failure(
+                    message=f"Unexpected error: {e} ({type(e).__name__})",
+                    context="query_history_extraction_unexpected_error",
+                )
+                return []
+
+    def populate_lineage_from_queries(self) -> None:
+        """Extract lineage from query history and add to SQL aggregator."""
+        if not self.config.include_query_lineage:
+            logger.info("Query-based lineage extraction disabled in config")
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py]: Prefer debug level.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py [hunk 18]
```diff
@@ -0,0 +1,335 @@
...
+            return
+
+        logger.info(
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/lineage.py]: Debug level is good for this log.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+        """Check if user has VIEW SERVER STATE permission for DMVs."""
+        return text("""
+            SELECT 
+                HAS_PERMS_BY_NAME(NULL, NULL, 'VIEW SERVER STATE') AS has_view_server_state
+        """)
+
+    @staticmethod
+    def get_query_history_from_query_store(
+        database: Optional[str],
+        limit: int,
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py:83]: Would be good to have validation for limit and min_calls. limit needs to > 0 and min_calls needs to be >= 0.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+            GROUP BY q.query_id, qt.query_sql_text
+            HAVING SUM(rs.count_executions) >= :min_calls
+            ORDER BY SUM(rs.avg_duration * rs.count_executions) DESC
+        """
+        )
+
+        return MSSQLQuery._finalize_query(query, exclude_patterns, limit, min_calls)
+
+    @staticmethod
+    def get_query_history_from_dmv(
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py:138]: This function and above query_store function is almost identical. Can we have one common function for identical part like:

```
  @staticmethod
  def _get_query_history_base(
      query_template: str,
      exclude_column_expr: str,
      limit: int,
      min_calls: int,
      exclude_patterns: Optional[List[str]],
  ) -> Tuple[TextClause, dict]:
      """Common logic for both Query Store and DMV methods."""
      exclude_clause = MSSQLQuery._build_exclude_clause(
          exclude_patterns, exclude_column_expr
      )

      final_query = query_template + exclude_clause + "..."  # Inject exclude clause

      return MSSQLQuery._finalize_query(final_query, exclude_patterns, limit, min_calls)
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+        Uses sys.dm_exec_cached_plans and sys.dm_exec_query_stats for cached queries.
+        This is the fallback method for older versions or when Query Store is disabled.
+
+        Note: DMVs only contain cached queries, so some historical queries may be missing.
+        """
+        exclude_clause = MSSQLQuery._build_exclude_clause(
+            exclude_patterns, "CAST(st.text AS NVARCHAR(MAX))"
+        )
+
+        query = (
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py]: Use f-string instead of string concatenation with +.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+        for i in range(len(exclude_patterns)):
+            condition = column_expr + " NOT LIKE :exclude_" + str(i)
+            conditions.append(condition)
+
+        return "AND " + " AND ".join(conditions)
+
+    @staticmethod
+    def _build_exclude_params(
+        exclude_patterns: Optional[List[str]], base_params: dict
+    ) -> dict:
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py]: Prefer specific return type(Dict[str, Union[int, str]]) than dict.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+                params[key] = pattern
+        return params
+
+    @staticmethod
+    def _finalize_query(
+        query_template: str,
+        exclude_patterns: Optional[List[str]],
+        limit: int,
+        min_calls: int,
+    ) -> Tuple[TextClause, dict]:
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py]: ```dict``` is vague.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py [hunk 24]
```diff
@@ -532,18 +834,16 @@ def _get_jobs_via_stored_procedures(
...
             }
 
-        # Now get job steps for each job, filtering by database
         for job_id, job_info in jobs_data.items():
             try:
-                # Get steps for this specific job
+                # job_id from sp_help_job output (database metadata)
                 steps_result = conn.execute(
-                    f"EXEC msdb.dbo.sp_help_jobstep @job_id = '{job_id}'"
+                    text("EXEC msdb.dbo.sp_help_jobstep @job_id = :job_id"),
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py:758]: Good change for safety.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py [hunk 25]
```diff
@@ -1486,44 +1513,52 @@ def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
...
+        Returns True if the table exists in discovered schemas, False if it's
+        a temp table, alias, or undiscovered table that should be filtered out.
 
-        Note: This method is called for each upstream table during lineage filtering.
-        If profiling shows this as a bottleneck, consider caching parsed URNs or
-        standardized names to reduce redundant processing.
+        Note: Uses instance-level caching to improve performance for repeated calls.
         """
+        if not hasattr(self, "_discovered_table_cache"):
+            self._discovered_table_cache: dict[str, bool] = {}
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py]: Would be good to initialize this in __init__.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py [hunk 26]
```diff
@@ -451,61 +652,166 @@ def _get_jobs(self, conn: Connection, db_name: str) -> Dict[str, Dict[str, Any]]
         """
         jobs: Dict[str, Dict[str, Any]] = {}
 
-        # Detect environment to choose optimal method first
         is_rds = self._detect_rds_environment(conn)
 
         if is_rds:
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py]: This logic is not introduced by this PR but it's getting complicated. Can we make it simpler by extracting exception handling to helper method to eliminate duplication and reduce nesting? For example:

```
def _get_jobs(self, conn: Connection, db_name: str) -> Dict[str, Dict[str, Any]]:
      is_rds = self._detect_rds_environment(conn)
      methods = (                                                                                                          
          [self._get_jobs_via_stored_procedures, self._get_jobs_via_direct_query]
          if is_rds                                                                                                        
          else [self._get_jobs_via_direct_query, self._get_jobs_via_stored_procedures]
      )

      return self._get_jobs_with_fallback(conn, db_name, methods, is_rds)

def _get_jobs_with_fallback(
    self,
    conn: Connection,
    db_name: str,
    methods: List[Callable],
    is_rds: bool,
) -> Dict[str, Dict[str, Any]]:
    """Try each method in order, handle all exception types once."""
    context_suffix = "rds" if is_rds else "on_prem"

    for method in methods:
        try:
            jobs = method(conn, db_name)
            logger.info("Retrieved jobs using %s", method.__name__)
            return jobs
        except (DatabaseError, OperationalError, ProgrammingError) as e:
            logger.warning("%s failed: %s", method.__name__, e)
        except (KeyError, TypeError) as e:
            logger.error("Job structure error %s: %s", method.__name__, e)
            self.report.failure(
                message=f"Job structure error: {e}",
                context=f"job_structure_error_{context_suffix}",
            )
        except Exception as e:
            logger.error("Unexpected error %s: %s", method.__name__, e, exc_info=True)
            self.report.failure(
                message=f"Unexpected error: {e}",
                context=f"job_extraction_error_{context_suffix}",
            )

    logger.error("All job retrieval methods failed for %s environment",
                 "RDS/managed" if is_rds else "on-premises")
    return {}
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py [hunk 27]
```diff
@@ -0,0 +1,335 @@
...
+                    f"SQL Server version {major_version} detected. "
+                    "Query Store requires SQL Server 2016+ (version 13). "
+                    "Falling back to DMV-based extraction."
+                )
+
+            return major_version
+
+        return None
+
+    def _check_query_store_available(self) -> bool:
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:119]: Some similarities between this and _check_dmv_permissions function.

We can make this simpler like the below (It's just an example):

```
  def _check_prerequisite(                                                                                                 
      self,                                                                                                                
      query,                                                                                                               
      condition: Callable[[dict], bool],
      success_msg: str,
      failure_msg: str,
  ) -> bool:
      """Execute check query and verify condition."""
      try:
          result = self.connection.execute(query)
          row = result.fetchone()

          if condition(row):
              logger.info(success_msg)
              return True
          else:
              logger.warning(failure_msg)
              return False
      except (DatabaseError, OperationalError, ProgrammingError) as e:
          logger.error("Check failed: %s", e)
          return False


  def _check_query_store_available(self) -> bool:
      return self._check_prerequisite(
          MSSQLQuery.check_query_store_enabled(),
          condition=lambda row: row and row["is_enabled"],
          success_msg="Query Store enabled - using Query Store",
          failure_msg="Query Store disabled - falling back to DMV",
      )


  def _check_dmv_permissions(self) -> bool:
      return self._check_prerequisite(
          MSSQLQuery.check_dmv_permissions(),
          condition=lambda row: row and row["has_view_server_state"],
          success_msg="VIEW SERVER STATE permission granted",
          failure_msg="Insufficient permissions - grant VIEW SERVER STATE",
      )
```

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+            ORDER BY SUM(rs.avg_duration * rs.count_executions) DESC
+        """
+        )
+
+        return MSSQLQuery._finalize_query(query, exclude_patterns, limit, min_calls)
+
+    @staticmethod
+    def get_query_history_from_dmv(
+        database: Optional[str],
+        limit: int,
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py:139]: Would be good to have validation for limit and min_calls. limit needs to > 0 and min_calls needs to be >= 0.

File: metadata-ingestion/tests/unit/test_mssql_query_lineage.py [hunk 29]
```diff
@@ -0,0 +1,1659 @@
...
+    assert any(
+        ("query" in msg.lower() and "lineage" in msg.lower())
+        or "extraction" in msg.lower()
+        or "unexpected error" in msg.lower()
+        for msg in failure_messages
+    )
+
+
+@patch("datahub.ingestion.source.sql.mssql.source.create_engine")
+def test_mssql_source_has_close_method(create_engine_mock):
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/tests/unit/test_mssql_query_lineage.py]: Testing hasattr and callable is testing Python language features, not behavior.

File: metadata-ingestion/tests/unit/test_mssql_query_lineage.py [hunk 30]
```diff
@@ -0,0 +1,1659 @@
...
+        "config": config,
+        "report": report,
+        "conn_mock": conn_mock,
+        "sql_aggregator_mock": sql_aggregator_mock,
+        "extractor": extractor,
+    }
+
+
+@patch("datahub.ingestion.source.sql.mssql.source.create_engine")
+def test_mssql_config_with_query_lineage(create_engine_mock):
```
Comment (kyungsoo-datahub) [inline metadata-ingestion/tests/unit/test_mssql_query_lineage.py]: I don't think we need this test. It's just testing that config parses and fields are set. Pydantic handles that automatically.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py (hunk 1 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py]: SQLServerConfig extends BasicSQLAlchemyConfig, which does NOT extend BaseUsageConfig or BaseTimeWindowConfig. Therefore, it's missing required fields. This is pre-exisiting in the code base. I have refactored maintaining the exisiting code. Let me know your suggests to handle this.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py (hunk 2 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py]: You're right. I followed the Postgres pattern without considering that SQL Server's Query Store is database-scoped by design. I'll remove it.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py (hunk 28 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py:139]: ACK

File: metadata-ingestion/tests/unit/test_mssql_query_lineage.py (hunk 29 above)
Comment (alokr-dhub) [reply metadata-ingestion/tests/unit/test_mssql_query_lineage.py]: ACK. wil remove it.

File: metadata-ingestion/tests/unit/test_mssql_query_lineage.py (hunk 30 above)
Comment (alokr-dhub) [reply metadata-ingestion/tests/unit/test_mssql_query_lineage.py]: removing it.

File: metadata-ingestion/docs/sources/mssql/mssql_pre.md (hunk 4 above)
Comment (alokr-dhub) [reply metadata-ingestion/docs/sources/mssql/mssql_pre.md]: DATA_FLUSH_INTERVAL and INTERVAL_LENGTH are SQL Server defaults per Microsoft's documentation. MAX_STORAGE_SIZE_MB is bumped from the default of 100MB to allow for larger query history retention

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py (hunk 6 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:1]: ACK

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py (hunk 19 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py:83]: fixed.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py (hunk 20 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query.py:138]: ACK.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py (hunk 1 above)
Comment (kyungsoo-datahub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/source.py]: This isn't pre-existing — no other source in the codebase passes a non-`BaseUsageConfig` config with `type: ignore`. I checked all 10 sources that use `usage_config`:

- **Postgres, Oracle, Redshift** — Config class inherits `BaseUsageConfig` directly (no type-ignore needed)
- **Teradata, Dremio, SqlQueries** — Passes a dedicated `.usage` sub-config that IS a `BaseUsageConfig`
- **Snowflake, BigQuery, ClickHouse** — Constructs a `BaseUsageConfig(...)` inline

The aggregator accesses `usage_config.bucket_duration`, `usage_config.buckets()`, and passes it to `UsageAggregator(config=...)` — so this **will crash at runtime** when `include_usage_statistics: true`.

Three options, ordered by preference:

**Option 1 — Inherit `BaseUsageConfig`** (matches Postgres/Oracle):
```python
class SQLServerConfig(BasicSQLAlchemyConfig, BaseUsageConfig):
```

**Option 2 — Construct inline** (matches Snowflake/BigQuery, keeps user config minimal):
```python
usage_config=BaseUsageConfig(
    bucket_duration=BucketDuration.DAY,
    start_time=...,
    end_time=...,
) if self.config.include_usage_statistics else None,
```

**Option 3 — Nested sub-config** (matches Teradata/Dremio):
```python
class SQLServerConfig(BasicSQLAlchemyConfig):
    usage: BaseUsageConfig = Field(default_factory=BaseUsageConfig)
```

I'd suggest Option 1 for simplicity, unless you want to avoid exposing `BaseUsageConfig` fields (like `top_n_queries`, `user_email_pattern`) to users — in that case Option 2 is cleanest.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py (hunk 27 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:119]: simplified the permission check logic

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py (hunk 11 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py]: ACK. removing the report failure

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py (hunk 16 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:228]: Added this for troubleshooting for different stages for query extraction separately for troubleshooting.

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py (hunk 27 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:119]: fixed this

File: metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py (hunk 16 above)
Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:228]: Let me know your suggestons

Comment (kyungsoo-datahub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:228]: I prefer removing them if they were not for features or not for metrics we're emitting. But it's trivial so it's up to you. If you see value, leave it. Remove it otherwise.

Comment (alokr-dhub) [reply metadata-ingestion/src/datahub/ingestion/source/sql/mssql/query_lineage_extractor.py:228]: ACK. let me remove it.

**Comment by rajatoss** [comment]
## Connector Tests Results

**All connector tests passed** for commit `67c5a4b`

[View full test logs →](https://github.com/acryldata/connector-tests/actions/runs/22350897429)

*Autogenerated by the connector-tests CI pipeline.*
